### DISCUSSION TITLE
vcpkg.json and versioning issue

### DISCUSSION THREAD

**Discussion by fer-marino** [microsoft/vcpkg#31166] [General]
hi everybody.
Please consider the following manifest:
```
{
  "name" : "eps",
  "version-string" : "1.0.0",
  "builtin-baseline" : "f32fc13dbef629b611fab9103d25e8d6c6ab9f6d",
  "dependencies" : [ {
    "name" : "xerces-c",
    "version>=" : "3.2.4"
  }, {
    "name" : "xalan-c",
    "version>=" : "1.12"
  } , {
    "name" : "omniorb",
    "version>=" : "4.3.0"
  }, {
    "name" : "coinutils",
    "version>=" : "2023-02-02"
  } , {
    "name" : "coin-or-clp",
    "version>=" : "2023-02-01"
  } , {
    "name" : "coin-or-osi",
    "version>=" : "2023-02-01"
  }  ]
}
```

The problem I'm having is that xalan-c has been delisted, so it is no more available in the latest baseline, while omniorb and coin* are available only on the recent baseline. How to resolve this? I have a feeling I should use overrides but I don't know how. Is it possible to define the latest baseline as a global settings and use the specific baseline only for xalan?

thank you!

PS I think delisting deprecated libs is not a good idea. Just show a warning...

**Comment by dg0yt** [comment]
Can you show the actual error message?
The delisting (in `versions/baseline.json`) was made in the belief that you can still refer to it (due to `versions/x-/xalan-c.json`).

**Comment by fer-marino** [comment]
If I run the manifest as is, with the latest baseline that included xalan, I get the following:

```
 error: Cannot resolve a minimum constraint for dependency coin-or-osi from coin-or-clp:x64-linux.
The dependency was not found in the baseline, indicating that the package did not exist at that time. This may be fixed by providing an explicit override version via the "overrides" field or by updating the baseline.
See `vcpkg help versioning` for more information.
```

If I update the baseline with the latest version (giving `vcpkg x-update-baseline`) I get the following:
```
error: the baseline does not contain an entry for port xalan-c
```

**Reply by dg0yt** [reply]
Thanks @fer-marino.

@BillyONeal In https://github.com/microsoft/vcpkg/pull/27401#issuecomment-1302626402, the statement was:

> Customers who want to still use this can pin an older version as older versions will still be present in the versions database.

This seems to be broken.

**Reply by fer-marino** [reply]
thanks. What now? shall I open an issue or add a comment to #27401 ?

**Reply by BillyONeal** [reply]
The example submitted uses constraints, rather than pinning (which would be in 'overrides').

I tried the following:

```json
{
  "name" : "eps",
  "version-string" : "1.0.0",
  "builtin-baseline" : "f32fc13dbef629b611fab9103d25e8d6c6ab9f6d",
  "dependencies" : [ {
    "name" : "xerces-c",
    "version>=" : "3.2.4"
  }, {
    "name" : "xalan-c",
    "version>=" : "1.12"
  } , {
    "name" : "omniorb",
    "version>=" : "4.3.0"
  }, {
    "name" : "coinutils",
    "version>=" : "2023-02-02"
  } , {
    "name" : "coin-or-clp"
  } , {
    "name" : "coin-or-osi"
  }  ],
  "overrides": [
    {"name": "coin-or-osi", "version": "2023-02-01"},
    {"name": "coin-or-clp", "version": "2023-02-01"}
  ]
}
```

which fails with:

```
error: version conflict on coinutils:x86-windows: eps:x86-windows required 2.11.4#1 but vcpkg could not compare it to 2023-02-02.
The two versions used incomparable schemes:
    "2023-02-02" was of scheme date
    "2.11.4#1" was of scheme string
This can be resolved by adding an explicit override to the preferred version, for example:
    "overrides": [
        { "name": "coinutils", "version": "2023-02-02" }
    ]
See `vcpkg help versioning` for more information.
```

after doing this a few times I got to:

```json
{
  "name" : "eps",
  "version-string" : "1.0.0",
  "builtin-baseline" : "f32fc13dbef629b611fab9103d25e8d6c6ab9f6d",
  "dependencies" : [ {
    "name" : "xerces-c",
    "version>=" : "3.2.4"
  }, {
    "name" : "xalan-c",
    "version>=" : "1.12"
  } , {
    "name" : "omniorb",
    "version>=" : "4.3.0"
  }, {
    "name" : "coinutils"
  } , {
    "name" : "coin-or-clp"
  } , {
    "name" : "coin-or-osi"
  }  ],
  "overrides": [
    {"name": "coin-or-osi", "version": "2023-02-01"},
    {"name": "coin-or-clp", "version": "2023-02-01"},
    {"name": "coin-or-buildtools", "version": "2023-02-02"},
    {"name": "coinutils", "version": "2023-02-02"},
    {"name": "coinutils", "version": "2023-02-02"},
    {"name": "glpk", "version": "5.0", "port-version": 1}
  ]
}
```

which sadly still said:

```
PS C:\Dev\test> ..\vcpkg\vcpkg.exe install --triplet x64-windows
error: the baseline does not contain an entry for port glpk
```

**Reply by BillyONeal** [reply]
@vicroms  @ras0219-msft  any ideas?

**Reply by dg0yt** [reply]
The key issue was getting xalan-c installed. IIUC it must be in `overrides` - and the issue is resolved?

The version scheme change issue is a different story, better found in https://github.com/microsoft/vcpkg/issues/31237.

**Reply by BillyONeal** [reply]
@dg0yt In my example glpk is in overrides but still is exploding. The version scheme results are expected. (version>= forces version comparisons so overrides being necessary to workaround that is by design)

**Reply by fer-marino** [reply]
thank you both for your help.

Honestly thought, this doesn't look effective. It should be the other way around: pinning xalan from an old baseline. The problem is that you would still face dependency hell if you are unlucky. So I repeat once more my original suggestion: don't delist deprecated libs but add a warning. NPM and Yarn do so, and they even add notices for vulnerabilities.

**Reply by vicroms** [reply]
Should be fixed by Billy's PR https://github.com/microsoft/vcpkg-tool/pull/1109

**Reply by fer-marino** [reply]
hi Vicroms,
can you please clarify how? I've tried to use the latest baseline and override only xalan but it didn't work

**Reply by vicroms** [reply]
The change hasn't made it to a tool release yet. You can try building the [latest vcpkg-tool from source](https://github.com/microsoft/vcpkg-tool).

**Reply by fer-marino** [reply]
hallo, I think this is merged. Still I couldn't get this work. Can you point me in the right direction please?

**Reply by BillyONeal** [reply]
@fer-marino You need to make sure the version you want is selected as an override. See the example in the original PR: https://github.com/microsoft/vcpkg-tool/pull/1109/files#diff-1b745c0d862e661dc64c96a28d79d62efa742f16a56ecb4132bbd8569223bf7dR14

**Reply by fer-marino** [reply]
this is my current manifest, I still get `error: the baseline does not contain an entry for port xalan-c`

```
{
  "name": "eps",
  "version-string": "3.x.x",
  "builtin-baseline": "d765306b074717dea8dc1c4723e1b025acb61c2d",
  "dependencies": [
    {
      "name": "xerces-c",
      "version>=": "3.2.4"
    },
    {
      "name": "xalan-c"
    },
    {
      "name": "pistache",
      "version>=": "2021-03-31#3"
    },
    {
      "name": "boost-log",
      "version>=": "1.82.0"
    },
    {
      "name": "boost-json",
      "version>=": "1.82.0"
    },
    {
      "name": "boost-asio",
      "version>=": "1.82.0"
    },
    {
      "name": "boost-thread",
      "version>=": "1.82.0"
    },
    {
      "name": "cppunit",
      "version>=": "1.15.1#3"
    },
    {
      "name": "antlr4",
      "version>=": "4.11.1"
    },
    {
      "name": "libmariadb",
      "version>=": "3.3.1#2"
    },
    {
      "name": "omniorb",
      "version>=": "4.3.0"
    },
    {
      "name": "coinutils",
      "version>=": "2023-02-02"
    },
    {
      "name": "coin-or-clp",
      "version>=": "2023-02-01"
    },
    {
      "name": "coin-or-osi",
      "version>=": "2023-02-01"
    }
  ],
  "overrides": [
    {
      "name": "xalan-c",
      "version": "1.12"
    }
  ]
}
````

**Reply by fer-marino** [reply]
I reinstalled vcpkg and it worked! Thanks for the fix and sorry for the noise
